### PR TITLE
feat: add request storage to support per-request context in ssr

### LINKED ISSUE #6151: feat: enable per-request context and DI for SSR
We need a way to isolate request-specific state so that Context and DI configuration doesn't get mixed up across requests from different users, as well as things like `window` and `document`. If possible, we should have a general purpose request storage solution in place with Context and DI layers on top of that.

### PR DESCRIPTION
**Pull request by EisenbergEffect**
# Pull Request

## 📖 Description

This PR creates some infrastructure to enable scoping certain "globals" to the request lifetime. By doing this, we are able to ensure that W3C Context is scoped pre request as well as DI.

The SSR example has been adjusted to use this new capability.

### 🎫 Issues

* Closes #6151

## 👩‍💻 Reviewer Notes

The main bit of code to review is in the `request-storage.ts` file. This file leverage's Node's `AsyncLocalStorage` API to provide request-scoped versions of `window` and `document`. This was built on top of a general-purpose `RequestStorage` solution that can be used for any request data, e.g. current user.

To see how this is used in an Express app with per-request Context, have a look at the SSR example.

## 📑 Test Plan

Basic tests were added. Some things are left untested due to the fact that I'm not quite sure how to test them in a way that doesn't effect the testing environment. Most of the code in these untested methods is very simple though, so the risk is very low.

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

## ⏭ Next Steps

n/a

### CODE REVIEW

**Review by nicholasrice — APPROVED** [review]
This all looks great! Can you add a section to the fast-ssr/README.md before this is merged though? I'd like to keep that doc up to date with the features even if it isn't *totally* fleshed out.

**Comment by EisenbergEffect** [comment]
> This all looks great! Can you add a section to the fast-ssr/README.md before this is merged though? I'd like to keep that doc up to date with the features even if it isn't _totally_ fleshed out.

Great idea! I'll add that before merging.

File: change/@microsoft-fast-foundation-c41dad2f-81e1-49b1-94ef-0b3115cc16d9.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "prerelease",
+  "comment": "feat: enable fallback container creation for DI location APIs",
```
Comment (awentzel) [inline change/@microsoft-fast-foundation-c41dad2f-81e1-49b1-94ef-0b3115cc16d9.json:3]: Should `DI` be spelled out? I'm not sure what it is.

File: change/@microsoft-fast-ssr-f38ce9ee-10a7-4f57-8e4f-ae7af978d36b.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "prerelease",
+  "comment": "feat: add request storage to support per-request context in ssr",
```
Comment (awentzel) [inline change/@microsoft-fast-ssr-f38ce9ee-10a7-4f57-8e4f-ae7af978d36b.json:3]: Just another note on abbreviations, I know SSR is server-side-rendering, but some people may not, so should we avoid abbreviating in things that end up in change logs.

**Review by awentzel — APPROVED** [review]

File: change/@microsoft-fast-foundation-c41dad2f-81e1-49b1-94ef-0b3115cc16d9.json (hunk 1 above)
Comment (awentzel) [reply change/@microsoft-fast-foundation-c41dad2f-81e1-49b1-94ef-0b3115cc16d9.json:3]: I guess it's dependency injection after finishing the review, though I still think we should avoid abbreviating except where using common industry acronyms.  It's common for those in the front-end, but maybe not for other layers or services of the stack.

Comment (EisenbergEffect) [reply change/@microsoft-fast-foundation-c41dad2f-81e1-49b1-94ef-0b3115cc16d9.json:3]: Good point here and below. I think I'm just used to having to keep these under the GitHub character limit for titles. But I guess that doesn't matter for our change files. Need to remember that. I'll update these before merging.

**Pull request merged**